### PR TITLE
move cards between tabs

### PR DESCRIPTION
**Pull request by npretto**
### Description

This implements milestone 1 of https://github.com/metabase/metabase/issues/34367
E2e test will be added in another PR, some tests are missing or will be refactored in another pr (see comments)

### How to verify

You should be able to move a dashcard between tabs

### Demo

https://github.com/metabase/metabase/assets/1914270/23b4d1d8-c52b-4b25-8644-6f66a98794dd

### CODE REVIEW

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.tsx [hunk 1]
```diff
@@ -1,17 +1,70 @@
...
+      {tabs.length > 1 && (
+        <TabChevronMenu tabs={tabs} onTabSelect={tabId => moveToTab(tabId)} />
+      )}
+
+      <VerticalDivider />
+    </>
+  );
+}
+
+type TabChevronMenuProps = {
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.tsx:45]: Personally I would split this out into a different file to keep one file <> one component

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import styled from "@emotion/styled";
+import type { MouseEventHandler } from "react";
+import { Icon } from "metabase/core/components/Icon";
+import { color } from "metabase/lib/colors";
+import type { AnchorProps } from "metabase/ui";
+import { Anchor } from "metabase/ui";
+
+export const VerticalDivider = styled.div`
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx:8]: You can use the Mantine `Divider` here 😄

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  align-self: stretch;
+  width: 1px;
+  background-color: ${color("border")};
+  padding: 5px 0px;
+`;
+
+export const ActionContainer = styled.div`
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx:15]: Instead of `div`, we can use `Box` (div), `Group` (flex row), or `Stack` (flex col), so we can take advantage of Mantine's style props. You still might need to use a styled component to add some of the unsupported props, but you can use the theme values here as well.


So for example, with this one we can do:

```
// In your JSX

<ActionContainer p="xs">
	... your stuff goes here ...
</ActionContainer>
```


```
export const ActionContainer = styled(Box)<BoxProps & HTMLAttributes<HTMLDivElement>>`

	&:hover {
		color: ${({theme}) => theme.colors.brand[1]};
	}

`
```

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  padding: 4px;
+  &:hover {
+    color: ${color("text-dark")};
+  }
+`;
+
+type TapButtonProps = AnchorProps & { onClick: MouseEventHandler };
+export const TabButton = styled(Anchor)<TapButtonProps>`
```
Comment (oisincoveney) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx:23]: With this one, you can set the props to
```
type TapButtonProps = AnchorProps & HTMLAttributes<HTMLAnchorElement>
```

Since the `Anchor` mantine component is based off off of `<a>`, we still get access to the React/HTML attributes here. This generally also applies to other Mantine components (i.e. `Box`<> `div`, `Text` <> `span/div`, `Button` <> `button`) - the docs go into a little more detail

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx (hunk 3 above)
Comment (npretto) [reply frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx:23]: Thanks, turns out that component was not used anymore 🙃

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.unit.spec.tsx [hunk 5]
```diff
@@ -1 +1,106 @@
...
+        ordered_tabs: tabs,
+      },
+    },
+  });
+
+  return renderWithProviders(<DashCardTabMenu dashCardId={123} />, {
+    storeInitialState: createMockState({ dashboard }),
+  });
+};
+
```
Comment (npretto) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.unit.spec.tsx:48]: all the tests needs to be re-done as the feature changed
I'll rewrite them after the testing plan has been reviewed

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import styled from "@emotion/styled";
+import type { HTMLAttributes } from "react";
+import { Menu } from "metabase/ui";
+import DashCardActionButton from "../DashCardActionButtons/DashCardActionButton";
+
+export const ItemWithMaxWidth = styled(Menu.Item)<
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.styled.tsx:6]: Overriding internal component styles is error-prone and should be always avoided. For this case you could do the following:
1. Pass a styled child to unchanged `Menu.Item` and set these properties on the child.
2. Pass `component` prop on `Menu.Item`

File: e2e/test/scenarios/dashboard/tabs.cy.spec.js [hunk 7]
```diff
@@ -81,6 +81,10 @@ describe("scenarios > dashboard > tabs", () => {
     });
   });
 
+  it("should allow moving dashcards between tabs", () => {
+    // TODO
```
Comment (npretto) [inline e2e/test/scenarios/dashboard/tabs.cy.spec.js:85]: will be added in another PR

File: frontend/src/metabase/dashboard/components/DashCard/DashCardActionButtons/DashCardActionButton/DashCardActionButton.tsx [hunk 8]
```diff
@@ -9,22 +9,22 @@ import {
...
-  children,
-  ...props
-}: Props) {
-  return (
-    <StyledAnchor {...props} data-metabase-event={analyticsEvent}>
-      <Tooltip tooltip={tooltip}>{children}</Tooltip>
-    </StyledAnchor>
-  );
-}
+const DashActionButton = React.forwardRef<HTMLAnchorElement, Props>(
```
Comment (npretto) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardActionButtons/DashCardActionButton/DashCardActionButton.tsx:16]: Mantine requires components used as Menu.Trigger to accept a ref: https://mantine.dev/core/menu/#required-ref-prop

File: frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.unit.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    });
+  });
+
+  describe("when selecting tab name from the menu", () => {
+    it("should move the card to the selected tab", async () => {
+      const { store } = setup({});
+
+      await clickOnTabMenuItem(TAB_3.name);
+
+      const storeDashcard = getDashCardById(store.getState(), DASHCARD.id);
```
Comment (npretto) [inline frontend/src/metabase/dashboard/components/DashCard/DashCardTabMenu/DashCardTabMenu.unit.spec.tsx:100]: All tests below this are checking the redux state, in a future PR I will move them to a upper component to avoid doing it

File: frontend/src/metabase/dashboard/actions/tabs.ts [hunk 10]
```diff
@@ -276,9 +300,35 @@ export const tabsReducer = createReducer<DashboardState>(
...
+          throw Error(
+            `MOVE_DASHCARD_TO_TAB was dispatched but dashId (${dashId}) is null`,
+          );
+        }
+        const dashCard = state.dashcards[dashCardId];
+
+        const { row, col } = getPositionForNewDashCard(
+          // @ts-expect-error: "Type instantiation is excessively deep and possibly infinite."
+          // Unfortunate problem that comes with using rtk draft state.
+          getExistingDashCards(state, dashId, destTabId),
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/tabs.ts:320]: @npretto just change the function definition

```
export function getExistingDashCards(
  dashboardState: Draft<DashboardState>,
```

I've verified locally that it works

**Review by ranquild — APPROVED** [review]
Works good, small comments. Please address before merging
